File: 2term/7.1. stack, list/7.1. stack,list/Stack.cs
namespace _7._1.stack_list
{
    /// <summary>
    /// Stack based on references.
    /// </summary>
    /// 
    public class Stack
    {
        /// <summary>
        /// Discription of Stack's element.
        /// </summary>
        private class StackElement
        {
            /// <summary>
            /// Meaning of the Stack's element.
            /// </summary>
            public int Value { get; set; }

            /// <summary>
            /// Field that point on the next element.
            /// </summary>
            public StackElement Next { get; set; }
        }

        /// <summary>
        /// Pointer of the first Stack's element which could exist or not.
        /// </summary>
        private StackElement head = null;

        /// <summary>
        /// Push value to a Stack.
        /// </summary>
        /// <param name="value">Value to be pushed.</param>
        public void Push(int value)
        {
            var newElement = new StackElement()
            {
                Next = head,
                Value = value
            };

            head = newElement;
        }

        /// <summary>
        /// Take Stack's top element.
        /// </summary>
        /// <returns></returns>
        public int Pop()
        {
            if (head == null)
            {
                return -1;
            }

            var temp = head.Value;
            head = head.Next;
            return temp;
        }

        /// <summary>
        /// Chek of existence elements in Stack.
        /// </summary>
        /// <returns></returns>
        public bool IsEmpty()
        {
            return head == null;
        }
    }
}

File: Qt(2)/2-2/stack.cpp
#include "stack.h"

Stack::Stack()
{
}

Stack::~Stack()
{
    while(!IsEmpty())
    {
        StackElement *tempElement = head;
        head = head->next;
        delete tempElement;
    }
}

void Stack::push(int value)
    {
        StackElement *newElement = new StackElement();
        newElement->value = value;
        newElement->next = head;

        head = newElement;
    }

int Stack::pop()
{
    if (head == nullptr)
    {
        return -1;
    }


    int temp = head->value;
    head = head->next;
    return temp;
}

bool Stack::isEmpty()
{
    return head == nullptr;
}
